simulator/func_sim/alu.h

```diff
@@ -233,6 +305,17 @@ struct ALU
     // Bit permutation
     template<typename I> static void grev( I* instr) { instr->v_dst = gen_reverse( instr->v_src1, shamt_v_src2<typename I::RegisterUInt>( instr)); }
 
+    template<typename I> static
+    void riscv_unshfl( I* instr)
+    {
+        auto dst_value = instr->v_src1;
+        constexpr size_t limit = log_bitwidth<decltype( instr->v_src1)> - 1;
+        for( size_t i = 0; i < limit; ++i)
+            if( ( instr->v_src2 >> i) & 1)
+                dst_value = bit_shuffle( dst_value, static_cast<decltype( instr->v_src1)>( 1 << i));
```

Comment:
If you have `size_t` as 2nd argument type, there will be no need of casting here (which is performed wrongly here anyway)